server/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -122,7 +122,13 @@ private RestRequest(
     }
 
     protected RestRequest(RestRequest other) {
+        this(other, Map.of());
+    }
+
+    private RestRequest(RestRequest other, Map<String, String> additionalParams) {
         assert other.parserConfig.restApiVersion().equals(other.restApiVersion);
+        this.params = other.params;
+        this.params.putAll(additionalParams);
```

Comment:
1. the implementation of the map is within the control of RestRequest - it is either a HashMap for the happy path or EMPTY_MAP for the path where we parsed incorrect headers. I think we would be able to add a testcase where RestRequest with incorrect header is enriched with a new parameter in serverless. 
2. ++
Agree that it would be nice to copy a map and perhaps a validation.

Reply:
> It feels off for a copy constructor. 

Indeed, this does violate the expectations for a copy constructor. I will fix, thanks !

> putAll overwrites existing entries.

I think that is a feature. For example, operator privs can change a true to false for a given param. It eliminates the need for special handling of restrictions if the response is already directly associated with a parameter. It is a flavor of injecting existing parameters to the rest request and I don't think we want to enforce that you can't inject existing parameters. 

> could also be helpful to prevent future APIs accidentally adding a parameter of the same name

I think that is also a feature.  You can introduce a general limited view as an optional response format, but allow operator privs to ensure it is always set to true.  

IMO if want to protect against param overrides we shouldn't  use params to express the restricted state, and instead use a `isRestricted` boolean ...but I prefer using params for the reasons above.